### ISSUE TITLE
ipfs cli output consistency

### ISSUE THREAD

**Issue by noffle** [ipfs/go-ipfs#2402]
Notes & discussion fodder on two `ipfs` usage points.

# subcommand discoverability

Some commands have subcommands. Some will output those subcommands when invoked without arguments. However, others will not, and must have `--help` provided:

```
$ ipfs
```
(shows all subcommands)

```
$ ipfs object
$ ipfs name
$ ipfs swarm
```
(shows all subcommands)

```
$ ipfs block
$ ipfs file
$ ipfs pin
$ ipfs repo
$ ipfs dht
$ ipfs diag
$ ipfs config
```
(no subcommands, just a terse summary)

Worse, some commands hide their subcommands! :(
```
$ ipfs bootstrap
(my peers)
```
```
$ ipfs bootstrap --help
(oh hey, there's 'list', 'add', and 'rm' subcommands!)
```

We can enhance discoverability greatly here by always outputting the subcommand listing when invoked.

Q: should `ipfs bootstrap` show peers, or show subcommand help? I lean toward the latter. It's a nice shortcut to avoid typing `ipfs bootstrap list`, but I'd rather have an alias on the first subcommand (e.g. `ipfs bs list`)

# show usage docs automatically when a command is invoked incorrectly

e.g.

```
stephen // ~ $ ipfs block get
Error: Argument 'key' is required

Use 'ipfs block get --help' for information about this command
```

99% of users probably ran this command this way because they don't know how to use it yet. It seems very likely that they will then immediately issue:

```
stephen // ~ $ ipfs block get --help

    ipfs block get <key> - Get a raw IPFS block.

ARGUMENTS:

    <key> - The base58 multihash of an existing block to get.

DESCRIPTION:

    'ipfs block get' is a plumbing command for retrieving raw ipfs blocks.
    It outputs to stdout, and <key> is a base58 encoded multihash.
```

Let's make this (usage output) the default when a bad invocation occurs.

---

cc @whyrusleeping, @diasdavid

**Comment by RichardLitt** [comment]
Also cc me. I think that `ipfs bootstrap` should show subcommands. Also agree about showing help directly without saying use help flag.

**Comment by noffle** [comment]
also https://github.com/ipfs/go-ipfs/issues/2408

**Comment by mnp** [comment]
I'd like to work on this.

**Comment by whyrusleeping** [comment]
Behaviour is as intended. Some commands have no need of arguments, if you wish to view the subcommands or helptext of a command use `--help`. 

Example, `ipfs swarm peers`. How would i invoke this command if just running it as is prints helptext? What argument would i pass it? `ipfs swarm peers list` is silly, might as well just rename the command to `ipfs swarm list-peers` at that point.

I also disagree with printing full helptext for every command failure, i don't want to scroll through hundreds of lines of text just because i messed something up. Although, i think its probably okay to print out a usage oneliner, like `USAGE: ipfs block get <key>`

**Comment by noffle** [comment]
> Example, ipfs swarm peers. How would i invoke this command if just running it as is prints helptext? 

Agreed: commands /wo subcommands should just run. *But*, ones that have subcommands should show them.

> I also disagree with printing full helptext for every command failure

Also agreed: we should show short-help instead of long-help here.
